mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp

```diff
@@ -2077,14 +2194,55 @@ void ConvertSYCLToLLVMPass::runOnOperation() {
   populateSYCLToLLVMConversionPatterns(converter, patterns);
   populateFuncToLLVMConversionPatterns(converter, patterns);
 
-  ConversionTarget target(*context);
-  target.addIllegalDialect<sycl::SYCLDialect>();
-  // TODO: connect wires to lower this operation.
-  target.addLegalOp<sycl::SYCLLocalIDOp>();
-  target.addLegalDialect<LLVM::LLVMDialect>();
-  target.addLegalDialect<arith::ArithDialect>();
+  populateVectorToLLVMConversionPatterns(converter, patterns);
+  arith::populateArithToLLVMConversionPatterns(converter, patterns);
+  populateSPIRVToLLVMTypeConversion(converter);
+  populateSPIRVToLLVMConversionPatterns(converter, patterns);
+  populateSPIRVToLLVMFunctionConversionPatterns(converter, patterns);
+
+  LLVMConversionTarget target(context);
+  target.addDynamicallyLegalDialect<sycl::SYCLDialect>(
+      [](Operation *op) { return isa<sycl::SYCLLocalIDOp>(op); });
+
+  const auto res = applyPartialConversion(module, target, std::move(patterns));
+
+  LLVM_DEBUG(llvm::dbgs() << "Module after LLVM lowering:\n"; module.dump(););
+
+  return res;
+}
+
+void ConvertSYCLToLLVMPass::cleanUnrealizedConversionCasts() {
+  LLVM_DEBUG(
+      llvm::dbgs() << "Reconciling UnrealizedConversionCasts operations...\n");
+
+  auto &context = getContext();
+  auto module = getOperation();
+
+  ConversionTarget target(context);
+  RewritePatternSet patterns(&context);
+
+  populateReconcileUnrealizedCastsPatterns(patterns);
+
+  if (applyPartialConversion(module, target, std::move(patterns)).failed()) {
+    LLVM_DEBUG(llvm::dbgs() << "WARNING: Could not remove every "
+                               "UnrealizedConversionCast operation\n");
+  }
+
+  LLVM_DEBUG(
+      llvm::dbgs()
+          << "Module after reconciling UnrealizedConversionCasts operations:\n";
+      module.dump(););
+}
 
-  target.addLegalOp<ModuleOp>();
-  if (failed(applyPartialConversion(module, target, std::move(patterns))))
-    signalPassFailure();
+void ConvertSYCLToLLVMPass::runOnOperation() {
+  constexpr int requiredRuns{2};
+  for (int i = 0; i < requiredRuns; ++i) {
+    if (convertToSPIRV().failed() || convertToLLVM().failed()) {
```

Comment:
IIUC, the second conversion to LLVM would still allow `SYCLLocalIDOp` as a legal operation, but we want to completely lower to LLVM dialect in that step. Should we enforce this by making the operation illegal on the second run?

Reply:
That's a good point. I've changed that. Thanks!